xarray/backends/common.py

```diff
@@ -249,11 +250,35 @@ def robust_getitem(array, key, catch=Exception, max_retries=6, initial_delay=500
 
 
 class BackendArray(NdimSizeLenMixin, indexing.ExplicitlyIndexed):
-    __slots__ = ()
-
     def get_duck_array(self, dtype: np.typing.DTypeLike = None):
         key = indexing.BasicIndexer((slice(None),) * self.ndim)
-        return self[key]  # type: ignore[index]
+        return self[key]  # type: ignore [index]
+
+
+class NewBackendArray(NdimSizeLenMixin, indexing.ExplicitlyIndexed):
+    __slots__ = ("indexing_support",)
+
+    def get_duck_array(self, dtype: np.typing.DTypeLike = None):
+        key = (slice(None),) * self.ndim
+        return self[key]  # type: ignore [index]
+
+    def _oindex_get(self, key: _OuterIndexerKey) -> Any:
+        raise NotImplementedError(
+            f"{self.__class__.__name__}._oindex_get method should be overridden"
+        )
+
+    def _vindex_get(self, key: _VectorizedIndexerKey) -> Any:
+        raise NotImplementedError(
+            f"{self.__class__.__name__}._vindex_get method should be overridden"
+        )
+
+    @property
+    def oindex(self) -> indexing.IndexCallable:
+        return indexing.IndexCallable(self._oindex_get)
+
+    @property
+    def vindex(self) -> indexing.IndexCallable:
+        return indexing.IndexCallable(self._vindex_get)
```

Comment:
as reminder the goal is to use `.oindex` and `.vindex` to be explicit about which indexing to use instead of `IndexingAdapter` subclasses.

I think this gets us nearly all the way to removing the special case for `ExplicitlyIndexed` in `as_compatible_data`